File: web/src/test/progressRepository.test.ts
import { describe, expect, it } from 'vitest';

import { createFitLogDb } from '@/db/client';
import { applyMigrations } from '@/db/migrate';
import { migrations } from '@/db/migrations';
import { seedCatalog } from '@/data/catalogSeed';
import { loadProgressSeries } from '@/data/progress';
import { addSet, deleteSet, finishSession, startSession } from '@/data/workout';
import { rangeFor } from '@/domain/progress';

import catalogSeed from '@shared/seed/catalog.json';

import { createNodeSqliteHarness } from './helpers/nodeSqlite';

type Db = ReturnType<typeof createFitLogDb>;

function first<T>(items: readonly T[]): T {
  const value = items[0];
  if (value === undefined) throw new Error('lista vacia');
  return value;
}

function createDatabase(): Db {
  const harness = createNodeSqliteHarness();
  applyMigrations(harness.runner, migrations);
  seedCatalog(harness.runner, catalogSeed, () => 1);
  return createFitLogDb(harness.client);
}

const exercise = first(catalogSeed.exercises.filter((item) => item.slug === 'press-banca-barra'));
const otherExercise = first(catalogSeed.exercises.filter((item) => item.slug === 'remo-barra'));

async function registerSession(
  db: Db,
  startedAt: number,
  sets: { weightKg: number | null; reps: number | null; isWarmup?: boolean; exerciseId?: string }[]
): Promise<string> {
  const session = await startSession(db, null, () => startedAt);
  for (const set of sets) {
    await addSet(
      db,
      {
        sessionId: session.id,
        exerciseId: set.exerciseId ?? exercise.id,
        weightKg: set.weightKg,
        reps: set.reps,
        rir: null,
        notes: null,
        isWarmup: set.isWarmup ?? false,
      },
      () => startedAt
    );
  }
  await finishSession(db, session.id, () => startedAt + 60_000);
  return session.id;
}

describe('loadProgressSeries', () => {
  it('construye un punto por sesion con las metricas del ejercicio', async () => {
    const db = createDatabase();
    const older = await registerSession(db, 1_000, [
      { weightKg: 100, reps: 8 },
      { weightKg: 40, reps: 10, isWarmup: true },
      { weightKg: 60, reps: 10, exerciseId: otherExercise.id },
    ]);
    const newer = await registerSession(db, 2_000, [{ weightKg: 105, reps: 5 }]);

    const points = await loadProgressSeries(db, exercise.id, null);

    expect(points.map((point) => point.sessionId)).toEqual([older, newer]);
    expect(points[0]?.workingSets).toBe(1);
    expect(points[0]?.maxWeightKg).toBe(100);
    expect(points[0]?.volumeKg).toBe(800);
    expect(points[0]?.bestOneRepMaxKg).toBeCloseTo(126.7, 9);
    expect(points[1]?.maxWeightKg).toBe(105);
    expect(points[1]?.bestOneRepMaxKg).toBeCloseTo(122.5, 9);
  });

  it('excluye series eliminadas', async () => {
    const db = createDatabase();
    const session = await startSession(db, null, () => 1_000);
    const kept = await addSet(db, {
      sessionId: session.id,
      exerciseId: exercise.id,
      weightKg: 100,
      reps: 8,
      rir: null,
      notes: null,
      isWarmup: false,
    });
    const removed = await addSet(db, {
      sessionId: session.id,
      exerciseId: exercise.id,
      weightKg: 120,
      reps: 1,
      rir: null,
      notes: null,
      isWarmup: false,
    });

    let points = await loadProgressSeries(db, exercise.id, null);
    expect(points[0]?.maxWeightKg).toBe(120);

    await deleteSet(db, removed.id, () => 2_000);
    points = await loadProgressSeries(db, exercise.id, null);
    expect(points[0]?.maxWeightKg).toBe(100);
    expect(points[0]?.workingSets).toBe(1);
    expect(kept.setIndex).toBe(1);
  });

  it('filtra por rango temporal', async () => {
    const db = createDatabase();
    await registerSession(db, 1_000, [{ weightKg: 100, reps: 8 }]);
    const recent = await registerSession(db, 2_000, [{ weightKg: 105, reps: 5 }]);

    const filtered = await loadProgressSeries(db, exercise.id, { fromMs: 1_500, toMs: 3_000 });
    expect(filtered.map((point) => point.sessionId)).toEqual([recent]);

    const all = await loadProgressSeries(db, exercise.id, rangeFor('all', 10_000));
    expect(all).toHaveLength(2);
  });

  it('devuelve lista vacia para un ejercicio sin series', async () => {
    const db = createDatabase();
    await registerSession(db, 1_000, [{ weightKg: 100, reps: 8 }]);

    const points = await loadProgressSeries(db, otherExercise.id, null);
    expect(points).toEqual([]);
  });
});

File: android/app/src/test/java/com/fitlog/app/ProgressRepositoryTest.kt
package com.fitlog.app

import android.content.Context
import androidx.room.Room
import androidx.test.core.app.ApplicationProvider
import com.fitlog.app.data.AddSetInput
import com.fitlog.app.data.CatalogRepository
import com.fitlog.app.data.FitLogDatabase
import com.fitlog.app.data.ProgressRepository
import com.fitlog.app.data.WorkoutRepository
import com.fitlog.app.domain.Progress
import kotlinx.coroutines.test.runTest
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34])
class ProgressRepositoryTest {

    private lateinit var database: FitLogDatabase
    private lateinit var progress: ProgressRepository
    private lateinit var workout: WorkoutRepository
    private lateinit var pressBanca: String
    private lateinit var remoBarra: String
    private var currentTime = 1_700_000_000_000L

    private val seedJson: String by lazy {
        val context = ApplicationProvider.getApplicationContext<Context>()
        context.assets.open("seed/catalog.json").bufferedReader().use { it.readText() }
    }

    @Before
    fun setUp() = runTest {
        database = Room.inMemoryDatabaseBuilder(
            ApplicationProvider.getApplicationContext(),
            FitLogDatabase::class.java,
        ).allowMainThreadQueries().build()

        val catalog = CatalogRepository(dao = database.catalogDao(), seedJsonProvider = { seedJson })
        catalog.ensureSeeded()

        workout = WorkoutRepository(
            dao = database.workoutDao(),
            routinesDao = database.routinesDao(),
            now = { currentTime },
        )
        progress = ProgressRepository(dao = database.progressDao())

        val exercises = catalog.loadCatalog().exercises
        pressBanca = exercises.first { it.slug == "press-banca-barra" }.id
        remoBarra = exercises.first { it.slug == "remo-barra" }.id
    }

    @After
    fun tearDown() {
        database.close()
    }

    private suspend fun registerSession(
        startedAt: Long,
        sets: List<AddSetInput>,
    ): String {
        currentTime = startedAt
        val session = workout.startSession(null)
        for (set in sets) {
            workout.addSet(set.copy(sessionId = session.id))
        }
        workout.finishSession(session.id)
        return session.id
    }

    private fun set(
        exerciseId: String,
        weightKg: Double?,
        reps: Int?,
        isWarmup: Boolean = false,
    ) = AddSetInput(
        sessionId = "",
        exerciseId = exerciseId,
        weightKg = weightKg,
        reps = reps,
        rir = null,
        notes = null,
        isWarmup = isWarmup,
    )

    @Test
    fun `construye un punto por sesion con las metricas del ejercicio`() = runTest {
        val older = registerSession(
            startedAt = 1_000,
            sets = listOf(
                set(pressBanca, 100.0, 8),
                set(pressBanca, 40.0, 10, isWarmup = true),
                set(remoBarra, 60.0, 10),
            ),
        )
        val newer = registerSession(startedAt = 2_000, sets = listOf(set(pressBanca, 105.0, 5)))

        val points = progress.series(pressBanca, null)

        assertEquals(listOf(older, newer), points.map { it.sessionId })
        assertEquals(1, points[0].workingSets)
        assertEquals(100.0, points[0].maxWeightKg, 1e-9)
        assertEquals(800.0, points[0].volumeKg, 1e-9)
        assertEquals(126.7, points[0].bestOneRepMaxKg, 1e-9)
        assertEquals(105.0, points[1].maxWeightKg, 1e-9)
        assertEquals(122.5, points[1].bestOneRepMaxKg, 1e-9)
    }

    @Test
    fun `excluye series eliminadas`() = runTest {
        val session = workout.startSession(null)
        val kept = workout.addSet(set(pressBanca, 100.0, 8).copy(sessionId = session.id))
        val removed = workout.addSet(set(pressBanca, 120.0, 1).copy(sessionId = session.id))

        var points = progress.series(pressBanca, null)
        assertEquals(120.0, points.first().maxWeightKg, 1e-9)

        workout.deleteSet(removed.id)

        points = progress.series(pressBanca, null)
        assertEquals(100.0, points.first().maxWeightKg, 1e-9)
        assertEquals(1, points.first().workingSets)
        assertEquals(1, kept.setIndex)
    }

    @Test
    fun `filtra por rango temporal`() = runTest {
        registerSession(startedAt = 1_000, sets = listOf(set(pressBanca, 100.0, 8)))
        val recent = registerSession(startedAt = 2_000, sets = listOf(set(pressBanca, 105.0, 5)))

        val filtered = progress.series(pressBanca, Progress.Range(fromMs = 1_500, toMs = 3_000))
        assertEquals(listOf(recent), filtered.map { it.sessionId })

        val all = progress.series(pressBanca, Progress.rangeFor(Progress.RangePreset.ALL, 10_000))
        assertEquals(2, all.size)
    }

    @Test
    fun `devuelve lista vacia para un ejercicio sin series`() = runTest {
        registerSession(startedAt = 1_000, sets = listOf(set(pressBanca, 100.0, 8)))

        assertTrue(progress.series(remoBarra, null).isEmpty())
    }
}
